pkg/server/cloud/gce/instances.go

```diff
@@ -0,0 +1,445 @@
+		accessConfig = nil
+	}
+	networkSpec := []*compute.NetworkInterface{
+		{
+			AccessConfigs: []*compute.AccessConfig{accessConfig},
+			AliasIpRanges: []*compute.AliasIpRange{
+				{
+					IpCidrRange: "/32",
+				},
+			},
+			Network:    networkURL,
+			Subnetwork: subNetworkURL,
+		},
+	}
+	return networkSpec
+}
+
+func (c *gceClient) createInstanceSpec(node *api.Node, metadata string) (*compute.Instance, error) {
+	md, err := base64.StdEncoding.DecodeString(metadata)
+	if err != nil {
+		return nil, util.WrapError(err, "Could not decode metadata string")
+	}
+	mds := string(md)
+	name := makeInstanceID(c.controllerID, node.Name)
+	diskType := c.getDiskTypeURL()
+	volSizeGiB := cloud.ToSaneVolumeSize(node.Spec.Resources.VolumeSize)
+	disks := c.getAttachedDiskSpec(true, int64(volSizeGiB), name, diskType, node.Spec.BootImage)
+	labels := c.getInstanceLabels(node.Name)
+	networkInterfaces := c.getInstanceNetworkSpec(node.Spec.Resources.PrivateIPOnly)
+	instanceType := c.getInstanceTypeURL(node.Spec.InstanceType)
+	spec := &compute.Instance{
+		Disks:             disks,
+		Labels:            labels,
+		MachineType:       instanceType,
+		Name:              name,
+		NetworkInterfaces: networkInterfaces,
+		Tags: &compute.Tags{
+			Items: c.bootSecurityGroupIDs,
+		},
+		Metadata: &compute.Metadata{
+			Items: []*compute.MetadataItems{
+				{
+					Key:   "user-data",
+					Value: &mds,
+				},
+			},
+		},
+	}
+	if node.Spec.Spot {
+		ar := false
+		spec.Scheduling = &compute.Scheduling{
+			AutomaticRestart:  &ar,
+			OnHostMaintenance: "TERMINATE",
+			Preemptible:       true,
+		}
+	}
+	return spec, nil
+}
+
+func (c *gceClient) StartNode(node *api.Node, metadata string) (*cloud.StartNodeResult, error) {
+	klog.V(2).Infof("Starting instance for node: %v", node)
+	spec, err := c.createInstanceSpec(node, metadata)
+	if err != nil {
+		return nil, err
+	}
+	klog.V(2).Infof("Starting node with security groups: %v subnet: '%s'",
+		c.bootSecurityGroupIDs, c.subnetName)
+	op, err := c.service.Instances.Insert(c.projectID, c.zone, spec).Do()
+	if err != nil {
+		return nil, util.WrapError(err, "startup error")
+	}
+	if err := c.waitOnOperation(op.Name, c.getZoneOperation); err != nil {
+		return nil, err
+	}
+	// Todo: catch and convert errors to notify us of
+	// out of capacity errors or invalid machine types
+	// see pkg/server/cloud/aws/instances.StartNode()
+	startResult := &cloud.StartNodeResult{
+		InstanceID:       spec.Name,
+		AvailabilityZone: c.zone,
+	}
+	return startResult, nil
+}
+
+func (c *gceClient) StartSpotNode(node *api.Node, metadata string) (*cloud.StartNodeResult, error) {
+	klog.V(2).Infof("Starting instance for node: %v", node)
+	spec, err := c.createInstanceSpec(node, metadata)
+	if err != nil {
+		return nil, err
+	}
+	klog.V(2).Infof("Starting node with security groups: %v subnet: '%s'",
+		c.bootSecurityGroupIDs, c.subnetName)
+	op, err := c.service.Instances.Insert(c.projectID, c.zone, spec).Do()
```

Comment:
Here too, how about using a context for timeouts?

Reply:
fixed in d7bc5a7